PickNikRobotics/moveit_pro_training_ws
# AprilTag detection Behavior and Objective

This PR adds a Behavior to detect AprilTag poses.

I made the AprilTag size a Behavior parameter / service request argument rather than a fixed parameter in the ROS 2 server so that it's easier to switch tag sizes between sim and hardware, and deal with the inevitable tag size tuning that is in our horizon :)

![image](https://github.com/PickNikRobotics/moveit_studio_training_ws/assets/4603398/d4cfdeca-c0f6-43d3-931e-34aff43a5ce3)

To test:
* Start up Studio with the `ur_gazebo_config` in this workspace
* Move to the "Behind" waypoint and tweak it a little bit to see the AprilTags on the table (these are IDs 1, 2, and 3)
* Look for the "Detect April Tag Pose" Objective and run it -- try both valid and invalid IDs
* (Bonus) If you want the OpenCV window to pop up, go to the `src/ur_gazebo_config/launch/sim/hardware_sim.launch.py` and modify the launch arg:

```python
    # AprilTag detection node
    apriltag_server_node = Node(
        package="apriltag_ros_python",
        executable="apriltag_detection_server",
        output="both",
        parameters=[
            {
                "visualize": True,  # <--- THIS ONE
                "apriltag_family": "tag36h11",
            }
        ],
    )
```
